File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.Stack;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];

        Stack<Integer> st = new Stack<>();

        for (int i = numbers.length - 1; i >= 0; i--) {
            answer[i] = -1; // 초기화

            while (!st.isEmpty()) {
                if (numbers[i] < st.peek()) {
                    answer[i] = st.peek();
                    break;
                } else {
                    st.pop();
                }
            }

            st.push(numbers[i]);
        }

        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.kt
import java.util.Stack

class Solution {
    fun solution(numbers: IntArray): IntArray {
        val answer = IntArray(numbers.size)

        val s = Stack<Int>()

        // 배열에서 마지막을 제외하고 반복한다.
        for (i in numbers.lastIndex downTo 0) {
            answer[i] = -1 // 초기화

            while (!s.isEmpty()) {
                if (numbers[i] < s.peek()) {
                    answer[i] = s.peek()
                    break
                } else {
                    s.pop()
                }
            }

            s.push(numbers[i])
        }

        return answer
    }
}